File: FBAudio-Android/app/src/main/java/com/fba/app/domain/LanguageFilter.kt
package com.fba.app.domain

import com.fba.app.domain.model.MenuNode
import com.fba.app.domain.model.SearchResult

/**
 * "English | All languages" support built on the markers FBA already maintains:
 *
 *  1. Curated menu labels carry a language suffix — "(deutsch)", "(en español)"…
 *     Bilingual entries ("deutsch | english") count as English.
 *  2. Places labels name the country — "(españa)", "(deutschland)"… Talks carry
 *     their centre, so talks from those centres can be hidden.
 *  3. Speakers listed under the Languages section, or marked per (1).
 *
 * The search API's language code is NOT used — it is unreliable (Spanish talks
 * are tagged "en"). India is deliberately left visible: talks there are in
 * English or Hindi and can't be told apart from the markers.
 */
object LanguageFilter {

    private val languageMarkers = listOf(
        "deutsch", "español", "espanol", "nederlands", "svenska", "français", "francais",
        "हिंदी", "polsku", "português", "portugues", "norsk", "русском", "suomeksi",
        "italiano", "auf deutsch", "en français",
    )

    private val nonEnglishCountryMarkers = listOf(
        "españa", "espana", "deutschland", "méxico", "mexico", "nederland", "vlaams",
        "france", "norge", "sverige", "brasil", "россия", "krakow", "kraków", "polska",
    )

    /** Does a curated menu label mark a non-English entry (and not a bilingual one)? */
    fun isNonEnglishLabel(label: String): Boolean {
        val l = label.lowercase()
        if (l.contains("english")) return false
        return languageMarkers.any { l.contains(it) }
    }

    /** Does a Places label name a non-English-speaking country? */
    fun isNonEnglishPlaceLabel(label: String): Boolean {
        val l = label.lowercase()
        return nonEnglishCountryMarkers.any { l.contains(it) }
    }

    /** Speaker names (lowercase) marked non-English in the People and Languages sections. */
    fun nonEnglishSpeakers(menu: List<MenuNode>): Set<String> {
        val out = mutableSetOf<String>()
        menu.firstOrNull { it.label.equals("people", ignoreCase = true) }?.children?.forEach { node ->
            if (isNonEnglishLabel(node.label)) speakerFromLink(node.link)?.let { out.add(it) }
        }
        menu.firstOrNull { it.label.equals("languages", ignoreCase = true) }?.children?.forEach { node ->
            speakerFromLink(node.link)?.let { out.add(it) }
        }
        return out
    }

    /** Centre names (lowercase) from Places entries in non-English-speaking countries. */
    fun nonEnglishCentres(menu: List<MenuNode>): Set<String> {
        val out = mutableSetOf<String>()
        fun visit(node: MenuNode) {
            if (isNonEnglishPlaceLabel(node.label)) {
                placeFromLink(node.link)?.let { out.add(it) }
                // "barcelona (españa)" → also match the city name itself
                out.add(node.label.substringBefore('(').trim().lowercase())
            }
            node.children.forEach { visit(it) }
        }
        menu.firstOrNull { it.label.equals("places", ignoreCase = true) }?.children?.forEach { visit(it) }
        menu.firstOrNull { it.label.equals("languages", ignoreCase = true) }?.children?.forEach { node ->
            placeFromLink(node.link)?.let { out.add(it) }
        }
        return out
    }

    /** Filter menu entries for the English-only setting. */
    fun filterMenu(nodes: List<MenuNode>, englishOnly: Boolean): List<MenuNode> {
        if (!englishOnly) return nodes
        return nodes.filterNot { isNonEnglishLabel(it.label) }
            .map { if (it.hasChildren) it.copy(children = filterMenu(it.children, true)) else it }
    }

    /** Filter talk/series items for the English-only setting. */
    fun filterItems(
        items: List<SearchResult>,
        englishOnly: Boolean,
        nonEnglishSpeakers: Set<String>,
        nonEnglishCentres: Set<String>,
    ): List<SearchResult> {
        if (!englishOnly) return items
        return items.filterNot { item ->
            val speaker = item.speaker.trim().lowercase()
            val centre = item.centre.trim().lowercase()
            (speaker.isNotBlank() && speaker in nonEnglishSpeakers) ||
                (centre.isNotBlank() && nonEnglishCentres.any { centre == it || centre.contains(it) }) ||
                isNonEnglishLabel(item.title)
        }
    }

    /** "/browse?s=Amalamati" → "amalamati" */
    fun speakerFromLink(link: String): String? =
        Regex("[?&]s=([^&#]+)").find(link)?.groupValues?.get(1)
            ?.let { java.net.URLDecoder.decode(it.replace("+", "%2B"), "UTF-8") }
            ?.replace('_', ' ')?.trim()?.lowercase()

    /** "/browse?p=Triratna_Barcelona" → "triratna barcelona" */
    fun placeFromLink(link: String): String? =
        Regex("[?&]p=([^&#]+)").find(link)?.groupValues?.get(1)
            ?.let { java.net.URLDecoder.decode(it.replace("+", "%2B"), "UTF-8") }
            ?.replace('_', ' ')?.trim()?.lowercase()
}

File: FBAudio-iOS/FBAudio/Data/LanguageFilter.swift
import Foundation

/// "English | All languages" support built on the markers FBA already maintains
/// (curated menu labels, the Languages section, Places country labels). The
/// search API's language code is NOT used — it is unreliable. India is left
/// visible: talks there are in English or Hindi. Mirrors Android's LanguageFilter.
enum LanguageFilter {

    private static let languageMarkers = [
        "deutsch", "español", "espanol", "nederlands", "svenska", "français", "francais",
        "हिंदी", "polsku", "português", "portugues", "norsk", "русском", "suomeksi",
        "italiano", "auf deutsch", "en français",
    ]

    private static let nonEnglishCountryMarkers = [
        "españa", "espana", "deutschland", "méxico", "mexico", "nederland", "vlaams",
        "france", "norge", "sverige", "brasil", "россия", "krakow", "kraków", "polska",
    ]

    /// Does a curated menu label mark a non-English entry (and not a bilingual one)?
    static func isNonEnglishLabel(_ label: String) -> Bool {
        let l = label.lowercased()
        if l.contains("english") { return false }
        return languageMarkers.contains { l.contains($0) }
    }

    /// Does a Places label name a non-English-speaking country?
    static func isNonEnglishPlaceLabel(_ label: String) -> Bool {
        let l = label.lowercased()
        return nonEnglishCountryMarkers.contains { l.contains($0) }
    }

    /// Speaker names (lowercase) marked non-English in the People and Languages sections.
    static func nonEnglishSpeakers(_ menu: [MenuNode]) -> Set<String> {
        var out = Set<String>()
        for node in SiteMenuParser.section(menu, "people")?.children ?? [] where isNonEnglishLabel(node.label) {
            if let s = speakerFromLink(node.link) { out.insert(s) }
        }
        for node in SiteMenuParser.section(menu, "languages")?.children ?? [] {
            if let s = speakerFromLink(node.link) { out.insert(s) }
        }
        return out
    }

    /// Centre names (lowercase) from Places entries in non-English-speaking countries.
    static func nonEnglishCentres(_ menu: [MenuNode]) -> Set<String> {
        var out = Set<String>()
        func visit(_ node: MenuNode) {
            if isNonEnglishPlaceLabel(node.label) {
                if let p = placeFromLink(node.link) { out.insert(p) }
                let city = node.label.split(separator: "(").first.map { String($0).trimmingCharacters(in: .whitespaces).lowercased() } ?? ""
                if !city.isEmpty { out.insert(city) }
            }
            node.children.forEach(visit)
        }
        (SiteMenuParser.section(menu, "places")?.children ?? []).forEach(visit)
        for node in SiteMenuParser.section(menu, "languages")?.children ?? [] {
            if let p = placeFromLink(node.link) { out.insert(p) }
        }
        return out
    }

    /// Filter menu entries for the English-only setting.
    static func filterMenu(_ nodes: [MenuNode], englishOnly: Bool) -> [MenuNode] {
        guard englishOnly else { return nodes }
        return nodes.filter { !isNonEnglishLabel($0.label) }.map { node in
            node.hasChildren
                ? MenuNode(label: node.label, link: node.link, om: node.om, children: filterMenu(node.children, englishOnly: true))
                : node
        }
    }

    /// Filter talk/series items for the English-only setting.
    static func filterItems(_ items: [SearchResult], englishOnly: Bool,
                            nonEnglishSpeakers: Set<String>, nonEnglishCentres: Set<String>) -> [SearchResult] {
        guard englishOnly else { return items }
        return items.filter { item in
            let speaker = item.speaker.trimmingCharacters(in: .whitespaces).lowercased()
            let centre = item.centre.trimmingCharacters(in: .whitespaces).lowercased()
            if !speaker.isEmpty && nonEnglishSpeakers.contains(speaker) { return false }
            if !centre.isEmpty && nonEnglishCentres.contains(where: { centre == $0 || centre.contains($0) }) { return false }
            if isNonEnglishLabel(item.title) { return false }
            return true
        }
    }

    /// "/browse?s=Amalamati" → "amalamati"
    static func speakerFromLink(_ link: String) -> String? {
        decodedQuery("s", link)
    }

    /// "/browse?p=Triratna_Barcelona" → "triratna barcelona"
    static func placeFromLink(_ link: String) -> String? {
        decodedQuery("p", link)
    }

    private static func decodedQuery(_ key: String, _ link: String) -> String? {
        guard let raw = ContentSource.queryValue(key, in: link) else { return nil }
        let withoutFragment = raw.split(separator: "#").first.map(String.init) ?? raw
        let decoded = withoutFragment.replacingOccurrences(of: "+", with: "%2B").removingPercentEncoding ?? withoutFragment
        return decoded.replacingOccurrences(of: "_", with: " ").trimmingCharacters(in: .whitespaces).lowercased()
    }
}
